### PR TITLE
Closes #6569: Add image dimension query to Saas request

### LINKED ISSUE #6569: 3.16 - When RUCSS & LCP/ATF are both enabled, URLs sent to the SaaS should have the wpr_imagedimensions=1 query string
**Context**
For SaaS visits to get the beacon script injected, the query string wpr_imagedimensions=1 must be present.
When both RUCSS & LCP/ATF are enabled, jobs sent to the SaaS are expected to do both the RUCSS optimization and trigger the beacon script.

**Issue**
On the current version of LCP/ATF, when RUCSS & LCP/ATF are enabled, URLs sent to the SaaS have this query string: ?nowprocket=1&no_optimize=1. wpr_imagedimensions=1 is missing so SaaS visits don't get the beacon script injected.

**Expected behavior**
- [x] When LCP/ATF and RUCSS are both deactivated, no URL must be sent to the SaaS
- [x] When LCP/ATF is on and RUCSS is off, URLs are sent to the SaaS as part of the pre-warm-up (can be triggered with "Clear Critical Images"). The query string must be `?wpr_imagedimensions=1%2F&nowprocket=1&no_optimize=1`.
- [x] When LCP/ATF is off and RUCSS is on, URLs are sent to the SaaS as part of the RUCSS process (can be triggered with "Clear Used CSS"). The query string must be `?nowprocket=1&no_optimize=1`.
- [ ] When LCP/ATF is on and RUCSS is on, URLs are sent to the SaaS as part of the RUCSS process (can be triggered with "Clear Used CSS"). The query string must be `?wpr_imagedimensions=1%2F&nowprocket=1&no_optimize=1`.

The last point is not working as expected.

**Acceptance Criteria**
- Replace the API_URL in the namespace WP_Rocket\Engine\Common\JobManager\APIHandler by the URL of our staging RUCSS server (you can ask on Slack to get it). This will allow you to check the URLs processed by the SaaS [on Metabase](https://data.wp-media.me/question#eyJkYXRhc2V0X3F1ZXJ5Ijp7ImRhdGFiYXNlIjoxNCwicXVlcnkiOnsic291cmNlLXRhYmxlIjo1NTAsIm9yZGVyLWJ5IjpbWyJkZXNjIixbImZpZWxkIiwxNDM4OSxudWxsXV1dfSwidHlwZSI6InF1ZXJ5In0sImRpc3BsYXkiOiJ0YWJsZSIsInZpc3VhbGl6YXRpb25fc2V0dGluZ3MiOnt9fQ==).
- Check the 3 last expected behaviors by setting the right LCP/ATF & RUCSS configuration, and triggering the warm-up or the RUCSS process with the "Clear xxx" button from the admin bar. Check the URLs processed by the SaaS.
- In all cases where RUCSS is on, the optimization_list parameters must be set to ['rucss'] in the request to the SaaS.
- If RUCSS is off and LCP/ATF is on, then optimization_list must be set to ''. (empty string).  

**Additional information**
The issue seems to be that `wpr_imagedimensions` is set in `add_to_atf_queue` which is a method only used by the warm-up LCP/ATF process, not the RUCSS one. Since `wpr_imagedimensions` should not be restricted to the warm-up, maybe it could be handled in the `add_to_queue` method of APIClient? We could either check there if lcp/Atf is enabled and add the wpr_imagedimensions if it is, or we could use a filter to maybe add query parameters, and have the LCP/ATF service subscribe to it to handle the addition of wpr_imagedimensions?

### PR DESCRIPTION
**Pull request by Khadreal**
# Description
Add image dimension as parameter that is sent to Saas
Fixes #6569

### User documentation

*Explain how this code impacts users.*

### Technical documentation

*Explain how this code works. Diagram & drawings are welcomed.*

## Type of change
- [X] Enhancement (non-breaking change which improves an existing functionality).

## New dependencies
*List any new dependencies that are required for this change.*

## Risks
*List possible performance & security issues or risks, and explain how they have been mitigated.*

# Checklists

## Feature validation

- [ ] I validated all the Acceptance Criteria. If possible, provide sreenshots or videos.
- [ ] I triggered all changed lines of code at least once without new errors/warnings/notices.
- [ ] I implemented built-in tests to cover the new/changed code.

## Documentation

- [ ] I prepared the user documentation for the feature/enhancement and shared it in the PR or the GitHub issue.
- [ ] The user documentation covers new/changed entry points (endpoints, WP hooks, configuration files, ...).
- [ ] I prepared the technical documentation if needed, and shared it in the PR or the GitHub issue.

## Code style
- [x] I wrote self-explanatory code about what it does.
- [x] I wrote comments to explain why it does it.
- [x] I named variables and functions explicitely.

## Observability
- [ ] I handled errors when needed.
- [ ] I wrote user-facing messages that are understandable and provide actionable feedbacks.
- [ ] I prepared ways to observe the implemented system (logs, data, etc.).

## Risks
- [ ] I explicitely mentioned performance risks in the PR.
- [ ] I explicitely mentioned security risks in the PR.

### CODE REVIEW

File: inc/Engine/Common/JobManager/APIHandler/APIClient.php [hunk 1]
```diff
@@ -32,15 +32,19 @@ class APIClient extends AbstractAPIClient implements LoggerAwareInterface {
...
 	public function add_to_queue( string $url, array $options ): array {
+		$url = add_query_arg(
+			[
+				'nowprocket'  => 1,
+				'no_optimize' => 1,
+			],
+			user_trailingslashit( $url )
+		);
+
+		$url = apply_filters( 'rocket_saas_api_queued_url', $url );
```
Comment (remyperona) [inline inc/Engine/Common/JobManager/APIHandler/APIClient.php:48]: Add filter docblock

File: inc/Engine/Media/AboveTheFold/WarmUp/Controller.php [hunk 2]
```diff
@@ -181,4 +181,21 @@ private function send_to_saas( $links ) {
 			usleep( $delay_between );
 		}
 	}
+
+	/**
+	 * Add wpr_imagedimensions to URL query
+	 *
+	 * @return string
+	 */
+	public function add_wpr_imagedimensions_query_arg(): string {
```
Comment (remyperona) [inline inc/Engine/Media/AboveTheFold/WarmUp/Controller.php]: this is missing the `$url` parameter

File: inc/Engine/Media/AboveTheFold/WarmUp/Controller.php [hunk 3]
```diff
@@ -181,4 +181,21 @@ private function send_to_saas( $links ) {
...
+		if ( ! $this->context->is_allowed() ) {
+			return '';
```
Comment (remyperona) [inline inc/Engine/Media/AboveTheFold/WarmUp/Controller.php]: should return the default `$url` instead

File: inc/Engine/Media/AboveTheFold/WarmUp/Subscriber.php [hunk 4]
```diff
@@ -57,4 +58,13 @@ public function warm_up_on_update( $new_version, $old_version ) {
...
 		$this->controller->warm_up();
 	}
+
+	/**
+	 * Add image dimensions query parameter to URL.
+	 *
+	 * @return string
+	 */
+	public function add_wpr_imagedimensions_query_arg(): string {
+		return $this->controller->add_wpr_imagedimensions_query_arg();
```
Comment (remyperona) [inline inc/Engine/Media/AboveTheFold/WarmUp/Subscriber.php]: missing the `$url` parameter

**Review by remyperona — CHANGES_REQUESTED** [review]

**Review by remyperona — APPROVED** [review]

**Comment by MathieuLamiot** [comment]
I manually validated the AC.
With RUCSS + LCPA/ATF, I can see both table filling too which means LCP/ATF is processed along side RUCSS.

**Pull request merged**